pkg/api/pluginproxy/ds_proxy.go

```diff
@@ -98,30 +98,52 @@ func (proxy *DataSourceProxy) HandleRequest() {
 		return
 	}
 
-	proxyErrorLogger := logger.New("userId", proxy.ctx.UserId, "orgId", proxy.ctx.OrgId, "uname", proxy.ctx.Login, "path", proxy.ctx.Req.URL.Path, "remote_addr", proxy.ctx.RemoteAddr(), "referer", proxy.ctx.Req.Referer())
-
-	reverseProxy := &httputil.ReverseProxy{
-		Director:      proxy.getDirector(),
-		FlushInterval: time.Millisecond * 200,
-		ErrorLog:      log.New(&logWrapper{logger: proxyErrorLogger}, "", 0),
-	}
+	proxyErrorLogger := logger.New("userId", proxy.ctx.UserId, "orgId", proxy.ctx.OrgId, "uname", proxy.ctx.Login,
+		"path", proxy.ctx.Req.URL.Path, "remote_addr", proxy.ctx.RemoteAddr(), "referer", proxy.ctx.Req.Referer())
 
 	transport, err := proxy.ds.GetHttpTransport()
 	if err != nil {
 		proxy.ctx.JsonApiErr(400, "Unable to load TLS certificate", err)
 		return
 	}
 
-	reverseProxy.Transport = &handleResponseTransport{
-		transport: transport,
+	reverseProxy := &httputil.ReverseProxy{
+		Director:      proxy.director,
+		FlushInterval: time.Millisecond * 200,
+		ErrorLog:      log.New(&logWrapper{logger: proxyErrorLogger}, "", 0),
+		Transport: &handleResponseTransport{
+			transport: transport,
+		},
+		ModifyResponse: func(resp *http.Response) error {
+			if resp.StatusCode == 401 {
+				// The data source rejected the request as unauthorized, convert to 400 (bad request)
+				body, err := ioutil.ReadAll(resp.Body)
+				if err != nil {
+					return fmt.Errorf("failed to read data source response body: %w", err)
+				}
+				_ = resp.Body.Close()
```

Comment:
I will have a look @papagian, can't say off the top of my head.

Reply:
I mean if the function returns in line 122 then `resp.Body` will not close.

Reply:
Thanks, will look later :)

Reply:
@papagian I don't think it's necessary to call `resp.Body.Close` because it should be the responsibility of the caller of `ModifyResponse`. The only reason `resp.Body.Close` is called in case of non-error is that we are replacing `resp` with a new `http.Response` object. Does that clear things up?

Reply:
ok, then